### PR TITLE
clients/js: add unit tests on worm CLI

### LINKED ISSUE #2596: clients/js: add tests
Currently, most if not all of the code in `clients/js` has no tests. We should add tests for each command to ensure future refactors don't break any functionality.

### PR DESCRIPTION
**Pull request by AlberErre**
Resolves #2596 

This PR adds unit tests  for `cmds` commands on worm CLI 🔒🚓 located on `client/ts/__tests__`

In order to run them, you need to navigate to `client/js` folder and run:

```
npm run test
```

Also, a `jest.config.json` file has been added to configure `jest` testing environment.

----

🚨 Note:  In order to test a command module, it needs to export all the module properties, following this [documentation](https://github.com/yargs/yargs/blob/main/docs/advanced.md#providing-a-command-module). 

There are some missing tests modules on this PR, such as `evm`, `aptos`, `generate`. These modules need refactor to be able to test them, as they don't expose all the required interfaces, in particular `exports.handler`. 

This refactor should be implemented in the future, as it's beyond the scope of this PR. 

## ✅ Update: 

All the `worm` commands are now covered with a test suite, checking all commands with their expected `positional arguments`, as well as checking if these commands are in sync with the auto-generated documentation (README.md).
In order to achieve this, we compare the worm CLI output expected by `worm <command>--help` pattern 👍🏻

Also, I've added functionality tests (to cover all cases dynamically) for the following commands:
  - `worm parse`
  - `worm recover`
  - `worm info chain-id`
  - `worm info rpc`
  - `worm info contract`

Once all these tests are executed, a HTML report is auto-generated to evaluate them if necessary 📝. This report is generated for every `npm run test` execution and is stored on `client/js/html-report/worm-cli-tests-report.html`.

Finally, a Github Actions (`.github/workflows/worm-cli.yml`) has been created to integrate all these tests & build process into wormhole pipelines 🚀 

Here is a recent Github Action pipeline I run successfully on my fork ✅:
https://github.com/AlberErre/wormhole/actions/runs/5337687871

### CODE REVIEW

**Comment by AlberErre** [comment]
@heyitaki @kcsongor @aadam-10 @evan-gray could you take a look at this? 👀 

feedback and suggestions are more than welcome! 🙏🏼 🚀

**Comment by AlberErre** [comment]
Hey @aadam-10 @heyitaki @evan-gray sorry to bother again, is there any chance to review this? 🙏🏼 

it's starting to diverge a lot from `main` branch 👀 🙏🏻

**Comment by heyitaki** [comment]
Hey @AlberErre, I'm hoping to merge in the #2885 stack first. There will likely be merge conflicts, but I'm happy to make those changes and push the PR through once that happens!

**Comment by heyitaki** [comment]
@AlberErre can you explain why the tests for the positional arguments are useful? I'm inclined to remove them... IMO as long as the functionality isn't broken and those corresponding tests pass it should be fine for a contributor to edit arg values/positions.

**Comment by AlberErre** [comment]
These commands are directly exposed to end users using worm CLI, therefore everyone relying on them will suffer breaking changes if we mess with them 👀.

It's similar to exposing a REST API. As long as new changes are extensible, we would be good to go and these tests won't break ✅. On the other hand, if the current public API is changed, these tests will break ❌, therefore notifying us that we are probably affecting end users with this change (in which case we will need to either avoid this change, or notify it on the README/CHANGELOG files). 

The point is that these commands are public, and therefore should be relatively stable. If these commands wouldn't be publicly exposed, we could indeed change them at our will, but it's not the case.

The idea is to provide a simple test suite emulating the flags and args expected on each command by end users, to figure out if we break them in future refactors 👀🏄‍♂️

**Comment by AlberErre** [comment]
now we can download the HTML test report directly from Github Actions 🎉, it's stored as an artifact on the pipeline. 

Here is an example: 
https://github.com/AlberErre/wormhole/actions/runs/5337687871

**Comment by AlberErre** [comment]
@aadam-10 @heyitaki @a5-pickle 

Just merged `main` to sync ✅, ready to review 👍🏻

**Review by evan-gray — CHANGES_REQUESTED** [review]
This is a lot of code which doesn't seem meaningfully more significant than this docs generation check that was added. Is this PR / design worth the maintenance overhead? https://github.com/wormhole-foundation/wormhole/blob/main/.github/workflows/build.yml#L256C11-L256C11

Tests that check the functionality could be good, but simple naming checking may be covered by the above.

**Comment by AlberErre** [comment]
Hey @evan-gray @aadam-10 just checked this new `make docs` command and looks good to auto-generate documentation for README files.

That said, I wouldn't use it for testing as they don't check the real implementation output ❌. 

👀 Rather, this is testing the output generated by `clients/js/src/doc.ts`, which is an abstraction layer we're implementing on top of `worm CLI` that could contain bugs or provide false positives.

🚨 Actually, this is what is happening right now on the `main` branch. As the README file is showing commands that doesn't exist such as `worm verify-vaa init-token-bridge` (https://github.com/wormhole-foundation/wormhole/blob/main/clients/js/README.md?plain=1#L941), among others.

I know these CLI tests could look like adding some unnecessary overhead, but they are more robust and secure ✅, as we're using the final executable (the same as end-users are using) to verify the implementation is working as expected (without any intermediate layer that could contain bugs).

**Comment by AlberErre** [comment]
@evan-gray just updated & deleted some tests as you suggested 👀

Also, I've added a more generic approach (and easy to maintain) to test command names with args. Now, we relay on README documentation generated by `clients/js/src/doc.ts` to tests these commands 🎉.

This way we still are comparing against the real implementation output ✅, but using README auto-generated docs to compare with. This ensures that documentation generation logic & worm CLI are in sync and produce the same output.

Some refactor has been needed to obtain a single source of commands to derive everything from. As a result, we now have a single source of truth in `src/cmds/index.ts`, regarding commands to be expected on both CLI usage & documentation.

This way, @kcsongor we don't need to maintain several arrays with the latest commands (e.g.  as commented here -> https://github.com/wormhole-foundation/wormhole/pull/3153#issuecomment-1613551798 for `doc.ts`), as now they're all defined on `src/cmds/index.ts`. 

Let me know what you think 👀, these tests are now far way easy to maintain, as well as still robust and secure ✅

**Comment by AlberErre** [comment]
@aadam-10 @heyitaki @a5-pickle @evan-gray @kcsongor

Just merged main to sync again ✅ , ready to review 🎉

File: clients/js/src/main.ts [hunk 1]
```diff
@@ -4,38 +4,12 @@ import { hideBin } from "yargs/helpers";
...
-  .command(parse)
-  .command(recover)
-  .command(submit)
-  .command(sui)
-  .command(transfer)
-  .command(verifyVaa)
-  .command(status)
+  // Build CLI commands dinamically from CLI_COMMAND_MODULES list
+  // Documentation about command hierarchy can be found here: https://github.com/yargs/yargs/blob/main/docs/advanced.md#example-command-hierarchy-using-indexmjs
+  .command(CLI_COMMAND_MODULES as YargsCommandModule[])
```
Comment (heyitaki) [inline clients/js/src/main.ts:13]: Why were you able to type `CLI_COMMAND_MODULES` directly as `YargsCommandModule[]` here, as opposed to using an intermediate `unknown` here? https://github.com/wormhole-foundation/wormhole/pull/2631/files#diff-e464885dd3e2f3456057320cb9e66fe189333a5eef5c2786cb027450ec52b445R11

**Review by heyitaki — CHANGES_REQUESTED** [review]

File: clients/js/src/main.ts (hunk 1 above)
Comment (heyitaki) [reply clients/js/src/main.ts:13]: I see this was answered, so marking as resolved. https://github.com/wormhole-foundation/wormhole/pull/3182/files#r1258347392

File: clients/js/__tests__/utils/getters.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+export type WormholeSDKChainName = keyof typeof CHAINS;
+
+export const getChains = () => {
+  return Object.keys(CHAINS) as WormholeSDKChainName[];
+};
+
+export type WormholeSDKNetwork = "mainnet" | "testnet" | "devnet";
+
+export const getNetworks = () => {
+  return ["mainnet", "testnet", "devnet"] as WormholeSDKNetwork[];
```
[the remark this answers is not in the document]
Comment (AlberErre) [reply clients/js/__tests__/utils/getters.ts]: You're 100% right ✅, just changed it to a const list

File: clients/js/__tests__/utils/getters.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { CHAINS } from "@certusone/wormhole-sdk/lib/esm/utils/consts";
+
+export type WormholeSDKChainName = keyof typeof CHAINS;
+
+export const getChains = () => {
+  return Object.keys(CHAINS) as WormholeSDKChainName[];
+};
+
+export type WormholeSDKNetwork = "mainnet" | "testnet" | "devnet";
```
[the remark this answers is not in the document]
Comment (AlberErre) [reply clients/js/__tests__/utils/getters.ts]: This was initially to type `getNetworks` method, just deleted this type as well ✅

File: clients/js/package.json [hunk 4]
```diff
@@ -65,8 +67,11 @@
     "@types/bn.js": "^5.1.0",
     "@types/bs58": "^4.0.1",
     "@types/node-fetch": "^2.6.3",
-    "@types/yargs": "^17.0.2",
+    "@types/yargs": "^17.0.24",
     "copy-dir": "^1.3.0",
+    "jest": "^29.5.0",
+    "jest-html-reporters": "^3.1.4",
```
[the remark this answers is not in the document]
Comment (AlberErre) [reply clients/js/package.json:73]: This is useful while tests are failing in the pipeline 🚨

By using this HTML report we can see which tests are failing (and why) more easily 🏄‍♂️. The current alternative is scrolling through CLI (Jest) error output, which is harder to read and reason about.

Regarding slimming our dependency list, this package is included in `devDependencies` 👀 . Therefore this dependency won't be included in the Worm CLI `main.js` executable output, and won't increase bundle size.

This dependency would only live during the lifespan of the Github Action, similar to what `Jest` dependency does.

File: clients/js/__tests__/recover.test.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+describe("worm recover", () => {
+  describe("check functionality", () => {
+    //mocks
+    const digest =
+      "0x99656f88302bda18573212d4812daeea7d39f8af695db1fbc4d99fd94f552606";
+    const signature =
+      "6da03c5e56cb15aeeceadc1e17a45753ab4dc0ec7bf6a75ca03143ed4a294f6f61bc3f478a457833e43084ecd7c985bf2f55a55f168aac0e030fc49e845e497101";
+
+    //expected output
+    const expectedAddress = "0x6FbEBc898F403E4773E95feB15E80C9A99c8348d";
```
[the remark this answers is not in the document]
Comment (AlberErre) [reply clients/js/__tests__/recover.test.ts:13]: Yes exactly, I launched this command manually with the mocked `digest` and `signature` values, and received this `expectedAddress` value in return.

I assumed the current implementation was correct, and therefore we can leverage on this result as a test case 👀

File: clients/js/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (heyitaki) [reply clients/js/package.json:73]: Hmm CLI output is usually enough for me, but can't hurt I guess 🤷

**Review by heyitaki — APPROVED** [review]

**Comment by AlberErre** [comment]
An extension of this PR (covering `worm submit` command) has been created here: https://github.com/wormhole-foundation/wormhole/pull/3229

File: clients/js/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlberErre) [reply clients/js/package.json:73]: @heyitaki just removed the HTML report generation on the next PR 👍 https://github.com/wormhole-foundation/wormhole/pull/3229

**Comment by evan-gray** [comment]
Closing this as some of these tests do not appear to be sufficiently meaningful or are already covered by other tests of the CLI. For example, the README.md is more efficiently covered by [the check](https://github.com/wormhole-foundation/wormhole/blob/main/.github/workflows/build.yml#L256) added in #3058

**Pull request closed without merging**